core/host.go

```diff
@@ -70,10 +70,9 @@ func (host *Host) Directory(ctx context.Context, dirPath string, p pipeline.Path
 		// synchronize concurrent filesyncs for the same path
 		llb.SharedKeyHint(localID),
 
-		// make the LLB stable so we can test invariants like:
-		//
-		//   workdir == directory(".")
-		llb.LocalUniqueID(localID),
+		// sync this dir from this session specifically, even if this ends up passed
+		// to a different session (e.g. a project container)
+		llb.SessionID(gw.BuildOpts().SessionID),
```

Comment:
Does this make `LocalUniqueID` redundant? iirc without it you could get different `DirectoryID` values for the same API call, so it was just added to make things more stable.

Reply:
The thing is, this does result in different DirectoryID values for every session. The only problem I thought of with this is cache invalidation in project commands when you pass a directory through to subcommands, e.g. here https://github.com/dagger/dagger/pull/5346/files#diff-44e0a27bc32810ce4627348a5ee1a73bba4101cc7c331105fd783946f32afd0fR13

I'm willing to accept that for the moment since I think having this ability to use project source code is really important and the downside is relatively minor at the moment.

I can think of a few hacks to patch that problem like passing the session ID to use in project containers through secrets, but I don't know if it's worth it yet. One more robust solution would be whole idea we talked about previously about making IDs based on the graphql query instead of including the LLB, with the full LLB cached in the graphql server.

Let me know if you are thinking of other problems this would potentially cause.

Reply:
I think in practice it's fine, I only added it before to make a test that was doing `worker == directory(".")` happy, so any issues were purely theoretical. We already don't promise that IDs are stable, the only thing that strictly matters is the LLB-level caching.

Reply:
> the only thing that strictly matters is the LLB-level caching

Agreed, and it turns out that if you don't set this, then buildkit will set it for you automatically and include it in the cache key for the op, so this shouldn't actually result in any new behavior on that level: https://github.com/sipsma/buildkit/blob/60d134bf7671e60bf2a2b9bc41a1d856e6ef396e/source/local/local.go#L70-L76